lib/puppet/util/at_fork/solaris.rb

```diff
@@ -0,0 +1,155 @@
+# contract is terminated.
+class Puppet::Util::AtFork::Solaris < Puppet::Util::AtFork
+  private
+
+  {
+    'libcontract.so.1' => [
+      #function name,            return value type, parameter types, ...
+      [:ct_ctl_abandon,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_activate,        Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+      [:ct_tmpl_clear,           Fiddle::TYPE_INT,  Fiddle::TYPE_INT],
+
+      [:ct_tmpl_set_informative, Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_tmpl_set_critical,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_param,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+      [:ct_pr_tmpl_set_fatal,    Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT],
+
+      [:ct_status_read,          Fiddle::TYPE_INT,  Fiddle::TYPE_INT, Fiddle::TYPE_INT, Fiddle::TYPE_VOIDP],
+
+      [:ct_status_get_id,        Fiddle::TYPE_INT,  Fiddle::TYPE_VOIDP],
+      [:ct_status_free,          Fiddle::TYPE_VOID, Fiddle::TYPE_VOIDP],
+    ],
+  }.each do |library, functions|
+    libhandle = Fiddle::Handle.new(library)
+
+    functions.each do |f|
+      define_method f[0], Fiddle::Function.new(libhandle[f[0].to_s], f[2..-1], f[1]).method(:call).to_proc
+    end
+  end
+
+  CTFS_PR_ROOT = File.join('', %w(system contract process))
+  CTFS_PR_TEMPLATE = File.join(CTFS_PR_ROOT, %q(template))
+  CTFS_PR_LATEST = File.join(CTFS_PR_ROOT, %q(latest))
+
+  CT_PR_PGRPONLY = 0x4
+  CT_PR_EV_HWERR = 0x20
+
+  CTD_COMMON = 0
+
+  def activate_new_contract_template()
+    begin
+      tmpl = File.open(CTFS_PR_TEMPLATE, File::RDWR)
+    rescue => detail
+      Puppet.log_exception(detail, _('Failed to open a new process contract template'))
+      return
+    end
+
+    begin
+      tmpl_fd = tmpl.fileno
+
+      raise if ct_pr_tmpl_set_param(tmpl_fd, CT_PR_PGRPONLY) != 0
+      raise if ct_pr_tmpl_set_fatal(tmpl_fd, CT_PR_EV_HWERR) != 0
+      raise if ct_tmpl_set_critical(tmpl_fd, 0) != 0
+      raise if ct_tmpl_set_informative(tmpl_fd, CT_PR_EV_HWERR) != 0
+
+      raise if ct_tmpl_activate(tmpl_fd) != 0
+    rescue
+      tmpl.close()
+      Puppet.err(_('Failed to configure / activate a new process contract template'))
+      return
+    end
+
+    @tmpl = tmpl
+  end
+
+  def deactivate_contract_template(where)
+    return if @tmpl.nil?
+
+    tmpl = @tmpl
+    @tmpl = nil
+
+    begin
+      raise if ct_tmpl_clear(tmpl.fileno) != 0
+    rescue
+      Puppet.fatal(_("Failed to deactivate process contract template in the #{where} process"))
+    ensure
+      tmpl.close()
+    end
+  end
+
+  def get_latest_child_contract_id()
+    begin
+      stat = File.open(CTFS_PR_LATEST, File::RDONLY)
+    rescue => detail
+      Puppet.log_exception(detail, _('Unable to access latest child process contract status'))
+      return nil
+    end
+
+    begin
+      stathdl = Fiddle::Pointer.new(0)
```

Comment:
It turns out you indeed can reset the `Fiddle::NULL` constant:
```ruby
require 'fiddle'
strtol = Fiddle::Function.new(Fiddle::Handle.new['strtol'], [Fiddle::TYPE_VOIDP, Fiddle::TYPE_VOIDP, Fiddle::TYPE_INT], Fiddle::TYPE_LONG)
a = '12xy'
b = Fiddle::NULL
puts "Fiddle::NULL address: #{Fiddle::NULL.to_i}"
puts "Calling strtol function ..."
c = strtol.call(Fiddle::Pointer.to_ptr(a), b.ref, 10)
puts "Fiddle::NULL address: #{Fiddle::NULL.to_i}"
puts "Integer value: #{c}"
puts "Unconverted string: #{b.to_s}"
```
which outputs:
```
Fiddle::NULL address: 0
Calling strtol function ...
Fiddle::NULL address: 140539989261538
Integer value: 12
Unconverted string: xy
```
Which is not something with what to do, so I'll stick with the current implementation.